2단계 - 수강신청(도메인 모델)

gibumgo opened the pull request:
리뷰 잘부탁드립니다.
구현 하면서 약간의 어색함이 있었지만, 이전에 배운 내용을 참고해서 구현해봤습니다.

요구사항에서 **"유료 강의는 수강생이 결제한 금액과 수강료가 일치할 때 수강 신청이 가능하다."**
이 부분이 payment 객체를 메소드로 받아오는 건지, user 객체에서 꺼내 오는 건지 정확히 이해하지 못해,
명료하게 구현하지 못한 것 같습니다. 확인 부탁드립니다. :sob:

gibumgo committed:
docs : 기능 구현 목록 작성

gibumgo committed:
feat : 강의 가격 정보 구분 기능 구현

gibumgo committed:
feat : 강의 기간을 가지는 기능 구현

gibumgo committed:
feat(courses) : 유료, 무료에 따른 강의 가격 입력 기능 구현

- ChargeTest 테스트 케이스 추가
- Charge 생성자 추가

gibumgo committed:
feat(courses) : 강의 상태 구분 기능 구현

gibumgo committed:
feat(courses) : 수강신청 기능 구현

- 최대 수강인원 제한 기능
- 강의 상태에 따른 수강 신청 가능 기능
- 강의 객체 및 강의 정보 객체 추가

gibumgo committed:
feat : 강의의 커버 이미지 생성 기능 구현

gibumgo committed:
feat : 수강 신청 기능 구현

- 무료 강의 최대 수강 인원 제한 없음.

gibumgo committed:
refactor(courses) : Students 일급컬렉션 객체 생성

- 수강 인원 중복 테스트 추가

gibumgo committed:
feat(courses) : 과정에 강의 리스트 추가

gibumgo committed:
style : 띄어쓰기 컨벤션 수정

gibumgo committed:
feat : 유료 강의 결제 기능 구현

nooose commented on src/test/java/nextstep/courses/domain/ChargeTest.java:
```diff
@@ -0,0 +1,25 @@
+        }).hasMessageContaining(PRICE_OF_FREE_SESSION_ERROR_MESSAGE);
+    }
+
+}
```
<img width="118" alt="image" src="https://github.com/next-step/java-lms/assets/36617783/d6479772-f000-4ecb-99df-294cbce36d36">

EOF 경고가 보이네요

이번 기수 리뷰어 최현구님의 [블로그](https://hyeon9mak.github.io/github-no-newline-at-a-end-of-file/)에서 잘 설명하고 있네요
참고해 주세요!

nooose commented on src/main/java/nextstep/courses/domain/ChargeStatus.java:
```diff
@@ -0,0 +1,18 @@
+
+public enum ChargeStatus {
+    FREE,
+    PAID;
```
강의 구분을 Charge에서 관리하는 방향을 선택하셨네요

지금은 강의 구분이 금액과 연관되어있지만
만약 나중에~ 요구사항이 변경되어 금액과 관련없는 다른 타입의 강의가 나온다면
어떻게 될까요?
ChargeStatus 값으로 계속 관리하기에는 약간 부자연스러울 것 같아요
참고해주세요~

nooose replied:
@gibumgo 
Strategy로 변경하셨네요
저의 의견은 금액(Charge) 타입보다는 강의(Session) 타입이 적절할 것 같다는 의견이었습니다. 😲
제가 드린 리뷰를 다시 읽어보니 제가 혼란을 드린 것 같네요...

nooose commented on src/main/java/nextstep/courses/domain/ChargeStatus.java:
```diff
@@ -0,0 +1,18 @@
+
+    private boolean isFree() {
+        return this == FREE;
+    }
```
👍

nooose commented on src/main/java/nextstep/courses/domain/Charge.java:
```diff
@@ -4,7 +4,12 @@ public class Charge {
+    public Charge(ChargeStatus chargeStatus) {
+        this.chargeStatus = chargeStatus;
+        this.price = 0;
+    }
```
```java
this(chargeStatus, 0);
```
주 생성자를 호출하면 좋을 것 같습니다.

nooose commented on src/main/java/nextstep/courses/domain/SessionStatus.java:
```diff
@@ -0,0 +1,7 @@
+    PREPARING,
+    ENROLLING,
+    FINISH;
+}
```
```java
    PREPARING("설명1"),
    ENROLLING("설명2"),
    FINISH("설명3");
```

언젠가는~ enum name이 변경될 수 있어요
네이밍이 마음에 들지 않아 등등.. 여러 이유가 있을 수 있죠
enum도 코드니까요. 그래서 저는 enum에도 의미를 담아주려고 하고 있습니다.

취향 차이가 있으니 참고만 해주세요~

nooose commented on src/main/java/nextstep/courses/domain/Enrollment.java:
```diff
@@ -0,0 +1,42 @@
+    public static final String MAX_CAPACITY_ERROR_MESSAGE = "강의 최대 수강 인원을 초과하였습니다.";
+    public static final String ENROLLMENT_ERROR_MESSAGE = "수강 신청 상태가 아닙니다.";
+    private final SessionStatus sessionStatus;
+    private final int capacity;
```
상수와 멤버변수간에는 개행으로 구분하는 것이 일반적인 컨벤션입니다.

nooose commented on src/main/java/nextstep/courses/domain/Enrollment.java:
```diff
@@ -0,0 +1,38 @@
+        this(sessionStatus, capacity, new Students());
+    }
+
+    public Enrollment(SessionStatus sessionStatus, int capacity, Students students) {
```
이 생성자가 필요할까요? 
테스트 코드에서도 사용하지 않는 생성자네요

`Enrollment(SessionStatus sessionStatus, int capacity)` 이 생성자에서 처리할 수 있지 않을까요?

nooose commented on src/main/java/nextstep/courses/domain/ImagePixel.java:
```diff
@@ -0,0 +1,49 @@
+    public static final String WIDTH_SIZE_ERROR_MESSAGE = "이미지의 너비는 %d pixel 이상이여야 합니다.";
+    public static final String IMAGE_RATIO_ERROR_MESSAGE = "이미지의 비율은 %d:%d여야 합니다.";
+    private int width;
+    private int height;
```
이 부분도 상수와 멤버변수를 구분하면 좋겠네요

nooose commented on src/main/java/nextstep/courses/domain/ImagePixel.java:
```diff
@@ -0,0 +1,49 @@
+    public static final int HEIGHT_RATIO = MIN_HEIGHT / 100;
+    public static final String HEIGHT_SIZE_ERROR_MESSAGE = "이미지의 높이는 %d pixel 이상이여야 합니다.";
+    public static final String WIDTH_SIZE_ERROR_MESSAGE = "이미지의 너비는 %d pixel 이상이여야 합니다.";
+    public static final String IMAGE_RATIO_ERROR_MESSAGE = "이미지의 비율은 %d:%d여야 합니다.";
```
다른 클래스에서 참조하는 값이 아니라면 private 접근제한자로 변경해볼 수 있겠네요

nooose commented on src/main/java/nextstep/courses/domain/ImagePixel.java:
```diff
@@ -0,0 +1,49 @@
+package nextstep.courses.domain;
+
+public class ImagePixel {
```
이름 정말 좋네요! 👍💯

nooose commented on src/main/java/nextstep/courses/domain/Session.java:
```diff
@@ -1,9 +1,49 @@
+    }
+
+    public static Session freeOf(String title, Long creatorId,
+                                 double imageSize, int width, int height, String imageType,
```
정적 팩토리 메소드 활용 좋습니다! 👍

nooose commented on src/test/java/nextstep/courses/domain/SessionTest.java:
```diff
@@ -0,0 +1,34 @@
+                800_000
+        );
+        paidSession.enroll(UserTest.JAVAJIGI);
+    }
```
검증 메소드가 없네요
등록이 잘 되었다는 것을 어떻게 확인하면 좋을까요?

nooose commented on src/main/java/nextstep/courses/domain/Students.java:
```diff
@@ -0,0 +1,30 @@
+    }
+    public boolean isFull(int capacity){
+        return capacity == students.size();
+    }
```
이 부분이 Students가 가져야할 메소드일까요?
Enrollment가 Students의 정보를 가지고 수강생이 오버했는지 판단할 수 있어보여요

nooose reviewed (changes requested):
안녕하세요 기범님
도메인 설계가 깔끔합니다. 👍
리뷰받고 넘어가면 좋을 것 같아요
재요청 부탁드립니다 🙇

> payment 객체를 메소드로 받아오는 건지, user 객체에서 꺼내 오는 건지 정확히 이해하지 못해,
명료하게 구현하지 못한 것 같습니다.

잘 구현하셨습니다! 👍

gibumgo committed:
refactor : 피드백 반영

- imagePixel 상수 접근제한자 변경
- 학생 수 구하는 기능 구현 (countEnrolledStudent() 메소드 추가)
- 학생 인원 관련 메소드 수정
- ChargeStatus, Charge 삭제 후, 강의 정책에 따른 전략 패턴 적용
- SessionStatus 이름 추가

kogibeom commented:
참고해서 봐주신 부분 또한 고민되었던 부분이었습니다. 실수도, 고민 되었던 부분도 세세히 짚어주셔서 정말 감사합니다.
리뷰어님께서 리뷰해주신 방향과 맞게 구현했을까요?

nooose commented on src/main/java/nextstep/courses/domain/Enrollment.java:
```diff
@@ -0,0 +1,54 @@
+import nextstep.users.domain.User;
+
+public class Enrollment {
+    private static final String NOT_FOUND_SESSION_ERROR_MESSAGE =
+            "수강 신청이 허용되지 않습니다. 결제 금액이 세션 가격과 일치하지 않습니다.";
+    public static final String MAX_CAPACITY_ERROR_MESSAGE = "강의 최대 수강 인원을 초과하였습니다.";
+    public static final String ENROLLMENT_ERROR_MESSAGE = "수강 신청 상태가 아닙니다.";
```
아래 두개는 public인데 어떤 차이가 있나요?

nooose commented on src/main/java/nextstep/courses/domain/Enrollment.java:
```diff
@@ -0,0 +1,54 @@
+import nextstep.users.domain.User;
+
+public class Enrollment {
+    private static final String NOT_FOUND_SESSION_ERROR_MESSAGE =
+            "수강 신청이 허용되지 않습니다. 결제 금액이 세션 가격과 일치하지 않습니다.";
+    public static final String MAX_CAPACITY_ERROR_MESSAGE = "강의 최대 수강 인원을 초과하였습니다.";
+    public static final String ENROLLMENT_ERROR_MESSAGE = "수강 신청 상태가 아닙니다.";
+    private final SessionStatus sessionStatus;
```
```suggestion
    private static final String NOT_FOUND_SESSION_ERROR_MESSAGE =
            "수강 신청이 허용되지 않습니다. 결제 금액이 세션 가격과 일치하지 않습니다.";
    public static final String MAX_CAPACITY_ERROR_MESSAGE = "강의 최대 수강 인원을 초과하였습니다.";
    public static final String ENROLLMENT_ERROR_MESSAGE = "수강 신청 상태가 아닙니다.";

    private final SessionStatus sessionStatus;
```

상수와 멤버변수는 개행으로 구분지어주세요

nooose commented on src/main/java/nextstep/courses/domain/ImagePixel.java:
```diff
@@ -0,0 +1,49 @@
+    public static final String HEIGHT_SIZE_ERROR_MESSAGE = "이미지의 높이는 %d pixel 이상이여야 합니다.";
+    public static final String WIDTH_SIZE_ERROR_MESSAGE = "이미지의 너비는 %d pixel 이상이여야 합니다.";
+    public static final String IMAGE_RATIO_ERROR_MESSAGE = "이미지의 비율은 %d:%d여야 합니다.";
+    private int width;
```
이 부분도 개행이 필요해보여요
다른 클래스도 확인 부탁드립니다.

nooose reviewed (approved):
반영 아주 잘 되었네요
놓친 부분은 제가 코멘트 남겼어요
다음 단계 진행해 주세요~

kogibeom commented:
금액(Charge) 타입 부분이 처음에 강의 타입인가? 생각했다 제가 복잡하게 생각했던 게 문제였네요ㅎㅎㅎㅎ
마지막 날임에도 리뷰 봐주셔서 정말 감사합니다. 사소하게 놓친 부분이 많아 아쉬움이 많네요 😭

nooose commented:
@kogibeom 
좋은 경험이 되었길 바래요
더 궁금한 부분이 있다면 DM 남겨주세요~
